### ISSUE TITLE
Cannot pickle '_thread.lock' object exception after DataArray transpose and copy operations from netCDF file.

### ISSUE THREAD

**Issue by sharkinsspatial** [pydata/xarray#8442]
### What is your issue?

I hit this issue while using [rioxarray](https://github.com/corteva/rioxarray) with a series of operations similar to those noted in this issue https://github.com/corteva/rioxarray/issues/614.  After looking through the `rioxarray` codebase a bit I was able to reproduce the issue with pure `xarray` operations.

If the Dataset is opened with the default `lock=True` settings,  [transposing](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.transpose.html) a DataArray's coordinates and then [copying](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.copy.html) the DataArray results in a `cannot pickle '_thread.lock' object` exception.

If the Dataset is opened with `lock=False`, no error is thrown.

This [sample notebook](https://nbviewer.org/gist/sharkinsspatial/c4512d2b8e53b6c42cd16c88f76e7325) reproduces the error.

This might be user error on my part, but it would be great to have some clarification on why `lock=False` is necessary here as my understanding was that this should only be necessary when using parallel write operations.

**Comment by sharkinsspatial** [comment]
Here is a locally reproducible MCVE.

```Python
import fsspec
import xarray as xr
import numpy as np

file_path = "test.nc"

ds = xr.Dataset(
    {
        'latitude': np.arange(10),
        'longitude': np.arange(10),
        'precip': (['latitude', 'longitude'], np.arange(100).reshape(10,10))

    }
)

ds = xr.open_dataset(file_path, engine="h5netcdf", decode_coords=True, decode_times=True)
da = ds["precip"]
da = da.transpose("longitude", "latitude", missing_dims="ignore")
da = da.copy()
```
Note that if `xr.open_dataset` is called with `lock=False` the `_io.BufferedReader error` is not thrown. 👍

**Comment by sharkinsspatial** [comment]
Apologies, I had not written the netCDF file out in the MCVE 🤦‍♂️, the example is updated now. I was able to produce the error in the environment below.

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.9.0 (default, Apr 14 2021, 14:07:04)
[Clang 12.0.0 (clang-1200.0.32.29)]
python-bits: 64
OS: Darwin
OS-release: 19.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: None

xarray: 2023.11.0
pandas: 2.1.3
numpy: 1.26.2
scipy: None
netCDF4: None
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: None
cftime: None
nc_time_axis: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
fsspec: 2023.12.1
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.2.2
pip: 23.3.1
conda: None
pytest: None
mypy: None
IPython: None
sphinx: None
</details>

**Comment by dcherian** [comment]
Mine too succeeds with `libhdf5: 1.14.2`, otherwise my versions of xarray, h5netcdf, h5py match yours.

**Comment by sharkinsspatial** [comment]
🤔 I upgraded to `libhdf5: 1.14.3` and was still able to reproduce.  To try and isolate any potential `h5netcdf` problems I also attempted the following with the default `netcdf4` engine and hit the same exception.
```
import xarray as xr
import numpy as np

file_path = "test.nc"

ds = xr.Dataset(
    {
        'latitude': np.arange(10),
        'longitude': np.arange(10),
        'precip': (['latitude', 'longitude'], np.arange(100).reshape(10, 10))
    }
)

ds.to_netcdf(file_path)

ds = xr.open_dataset(file_path, decode_coords=True, decode_times=True)
da = ds["precip"]
da = da.transpose("longitude", "latitude", missing_dims="ignore")
da = da.copy()
```
I'm going to ask a few colleagues to try and replicate to see if this is something peculiar to my environment.

**Comment by sharkinsspatial** [comment]
My colleague was also able to reproduce with the ☝️ `netcdf4` engine code and the following enviornment.

**Comment by zmoon** [comment]
No idea if it has the same underlying cause (I'm not transposing), but I do have a situation that used to work but now[^1] gives this same `cannot pickle '_thread.lock' object` error. I'll have to see if I can make it into a minimal example. Tried downgrading some things in my environment to no avail.

[^1]: First noticed a month or two ago I think.

**Comment by zmoon** [comment]
I was able to reproduce error in [OP's above example](https://github.com/pydata/xarray/issues/8442#issuecomment-1841971760). Similar to one of [my experiments](https://github.com/pydata/xarray/issues/8442#issuecomment-1853283162), error averted if adding a new variable to the Dataset (e.g. `ds["asdf"] = xr.zeros_like(ds.precip)`) before the transpose line.

**Comment by kmuehlbauer** [comment]
@zmoon Thanks for this MCVE! I can't reproduce the error, though. Also the MCVE in https://github.com/pydata/xarray/issues/8442#issuecomment-1841971760 works nicely (details below).

Does it still fail if environments are created from scratch or on other systems? It looks like linux itself is not affected, only MacOSX and WSL?

<details><summary>Versions</summary>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 4.19.0-22-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8
LOCALE: ('de_DE', 'UTF-8')
libhdf5: 1.14.2
libnetcdf: 4.9.2

xarray: 2023.12.0
pandas: 2.1.3
numpy: 1.26.2
scipy: 1.11.4
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: None
cftime: 1.6.3
nc_time_axis: None
iris: None
bottleneck: None
dask: 2023.11.0
distributed: 2023.11.0
matplotlib: 3.8.2
cartopy: 0.22.0
seaborn: None
numbagg: None
fsspec: 2023.10.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.2.2
pip: 23.3.1
conda: None
pytest: None
mypy: None
IPython: 8.18.1
sphinx: None
</details>

**Comment by kmuehlbauer** [comment]
OK, here we go, I've taken `dask` out of the loop in a fresh env and can now reproduce both MCVE.

<details><summary>Versions</summary>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 5.14.21-150500.55.19-default
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8
LOCALE: ('de_DE', 'UTF-8')
libhdf5: 1.14.3
libnetcdf: 4.9.2

xarray: 2023.12.0
pandas: 2.1.4
numpy: 1.26.2
scipy: None
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: None
cftime: 1.6.3
nc_time_axis: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
fsspec: None
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.2.2
pip: 23.3.1
conda: None
pytest: None
mypy: None
IPython: 8.18.1
sphinx: None
```
</details>

**Comment by zmoon** [comment]
@kmuehlbauer I experienced the error on Windows as well as WSL.

I tried a fresh env on Linux and still got the error 🤷 

<details><summary>Versions</summary>

```
mamba create -n test-lock python=3.11 xarray pooch netcdf4 h5netcdf joblib
```
```
INSTALLED VERSIONS
------------------
commit: None
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-957.27.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.3
libnetcdf: 4.9.2

xarray: 2023.12.0
pandas: 2.1.4
numpy: 1.26.2
scipy: None
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: None
cftime: 1.6.3
nc_time_axis: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
fsspec: None
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.2.2
pip: 23.3.1
conda: None
pytest: None
mypy: None
IPython: None
sphinx: None
```
</details>

**Comment by kmuehlbauer** [comment]
There has been some refactoring lately involving dask and other ChunkManagers. Not sure, if this has anything to do with it, but maybe @TomNicholas has more insight here.

**Comment by TomNicholas** [comment]
I don't really see why this should have anything to do with it... I guess it's not impossible that somehow some dask `lock` argument is now getting lost, but I suggest that if we can now reproduce the error someone should do a `git-bisect` to find out which commit caused the regression.

**Comment by kmuehlbauer** [comment]
Yes, thanks @TomNicholas for looking into this. Will try to bisect this.

**Comment by kmuehlbauer** [comment]
@zmoon @sharkinsspatial Did this ever work for you? I've a hard time finding a working commit. I've checked several versions back to 0.17.0 without success. Also the other involved dependencies (hdf5, netcdf-c, netcdf4-python, h5py, pandas) would be good to know to recreate a working environment.

**Comment by zmoon** [comment]
@kmuehlbauer for me I don't have the environment anymore, but I suspect I probably had dask installed in it.

**Comment by kmuehlbauer** [comment]
### TL;DR:

The current default of `xr.open_dataset` (netcdf4/h5netcdf) uses lazy loading which uses `threading.Lock` as default locking mechanism if `dask` is not available. The object cannot be pickled and after some computations (here `.transpose`) also not (deep)-copied. The only way around is to either explicitly use `lock=False` when opening files or do a `.load()` or `.compute()` before pickle/copy.

### Inspection:

Using the MCVE given here https://github.com/pydata/xarray/issues/8442#issuecomment-1841971760 I checked the types of the underlying array and how this works for transposing or not:

- `cache=True` in `open_dataset` (default)
    - no transpose
        - before copy: `<class 'xarray.core.indexing.MemoryCachedArray'>`
        - after copy: `<class 'xarray.core.indexing.MemoryCachedArray'>`
        - trying to pickle raises `TypeError: cannot pickle '_thread.lock' object` in pickle
    - with transpose
        - before transpose: `<class 'xarray.core.indexing.MemoryCachedArray'>`
        - after transpose: `<class 'xarray.core.indexing.LazilyVectorizedIndexedArray'>`
        - trying to copy raises: `TypeError: cannot pickle '_thread.lock' object` in deepcopy
        
- `cache=False` in `open_dataset`
    - no transpose
        - before copy: `<class 'xarray.core.indexing.CopyOnWriteArray'>`
        - after copy: `<class 'xarray.core.indexing.CopyOnWriteArray'>`
        - trying to pickle raises `TypeError: cannot pickle '_thread.lock' object` in pickle
    - with transpose
        - before transpose: `<class 'xarray.core.indexing.CopyOnWriteArray'>`
        - after transpose: `<class 'xarray.core.indexing.LazilyVectorizedIndexedArray'>`
        - trying to copy raises: `TypeError: cannot pickle '_thread.lock' object` in deepcopy

Reading with `netcdf4` and `h5netcdf` backends the data is wrapped in xarray's lazy classes See https://docs.xarray.dev/en/stable/user-guide/io.html#netcdf:

> Data is _always_ loaded lazily from netCDF files. You can manipulate, slice and subset Dataset and DataArray objects, and no array values are loaded into memory until you try to perform some sort of actual computation.

and further:

> Xarray’s lazy loading of remote or on-disk datasets is often but not always desirable. Before performing computationally intense operations, it is often a good idea to load a Dataset (or DataArray) entirely into memory by invoking the [Dataset.load()](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.load.html#xarray.Dataset.load) method.

There is also a mention for Pickle:

https://docs.xarray.dev/en/stable/user-guide/io.html#pickle

> When pickling an object opened from a NetCDF file, the pickle file will contain a reference to the file on disk. If you want to store the actual array values, load it into memory first with [Dataset.load()](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.load.html#xarray.Dataset.load) or [Dataset.compute()](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.compute.html#xarray.Dataset.compute).

### What to do?

The pickle issue might not be the big problem as the user is advised to load/compute before. But the copy-issue should be resolved somehow. Unfortunately I do not have an immediate solution to this. @pydata/xarray any ideas?

**Comment by max-sixty** [comment]
(brief message to say thanks a lot @kmuehlbauer for the excellent summary)

**Comment by shoyer** [comment]
I believe the issue are these two default locks for HDF5 and NetCDFC:
https://github.com/pydata/xarray/blob/2971994ef1dd67f44fe59e846c62b47e1e5b240b/xarray/backends/locks.py#L18

Probably the easiest way to handle this is to fork the code for SerializableLock from dask. It isn't very complicated:
https://github.com/dask/dask/blob/6f2100847e2042d459534294531e8884bef13a99/dask/utils.py#L1160

**Comment by kmuehlbauer** [comment]
Thanks @shoyer!
